### ISSUE TITLE
Should we standardize the anatomy of `<input type=range>`?

### ISSUE THREAD

**Issue by mfreed7** [WICG/open-ui#297]
First, I'm a bit unclear on Open-UI vs. WhatWG HTML for this. This issue is about standardizing the pseudo element parts of the Web Platform `<input type=range>`. Does this belong here? Or in WHATWG? I do see that there is a [research page](https://open-ui.org/components/slider.research) for "slider" which is the Open UI concept implemented by `<input type=range>`.

## Problem
1. No healthy way to style the filled/progress amount of the slider
    1. CSS only solutions are bound to the size of the track based on the only cross browser solution possible 
        1. [example from native-ui](https://app.native-elements.dev/editor/elements/range) (about as beautiful as they get)
        1. [example](https://codepen.io/argyleink/pen/WNoxjLL): since this design called for the thumb to be larger than the track, CSS only can't show the "filled" amount
    1. current prefixed selectors stomp each other, so developers are forced to duplicate styles
        1. Josh Comeau sharing their learnings about this oddity: [tweet](https://twitter.com/JoshWComeau/status/1359213597331763200?s=20)
    1. Javascript solution runs on page load and watches for user input to update a custom property
        1. [example](https://codepen.io/argyleink/pen/OJbXOvY)
    1. Firefox has `::-moz-range-progress`
        1. [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/::-moz-range-progress)
    1. we've recently added support for `accent-color` which will tint this fill range state, but still can't target it with a selector and give it a gradient background
2. Partial support of <datalist> [ticks](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/range#A_range_control_with_hash_marks:~:text=A%20range%20control%20with%20hash%20marks,-This) and labels
    1. [vertical] completely unsupported with ticks
        1. vertical tucked behind a prefix `-webkit-appearance: slider-vertical;` why?
    1. labels missing
    1. missing selectors for styling ticks / labels

## Tasks:
1. Separate `track` from `progress` and enable `::-webkit-slider-progress-track` as a style target for CSS
2. Complete current ticks implementation to work with vertical
    1. unprefix the selectors
    1. support attribute
3. Add labels to our slider implementation
4. Help spec a way to style the labels and ticks

## Resources
- Range ticks/hashes/labels: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/range#A_range_control_with_hash_marks 
- HTML 5.1 spec: https://www.w3.org/TR/html51/sec-forms.html#range-state-typerange 

--> Full credit due here to @argyleink

**Comment by una** [comment]
1000% agree we need standardization on this. Thanks for opening this, @mfreed7

**Comment by gregwhitworth** [comment]
@mfreed7  I've started the work on this in the past: https://open-ui.org/components/slider.research.parts

Also, I outlined to @emilio what the next steps are here because Firefox wants this standardized too: https://github.com/w3c/csswg-drafts/issues/4410#issuecomment-783022435

**Comment by gregwhitworth** [comment]
@mfreed7 @emilio circling back on this as @una and I were just discussing this issue. Is the primary goal here to solely standardize the named parts? Because that can be done quickly - there will still be a LOT of interop styling issues without going further but that would at least remove the need for numerous prefixed props.

**Comment by mfreed7** [comment]
Thanks for the pointer to https://github.com/w3c/csswg-drafts/issues/4410#issuecomment-783022435, I had seen that but hadn't read it in detail. I'm in favor of trying to do what you lay out in that comment - i.e. try to really "fix" the `<input type=range>`, not just do the quick fix. Assuming there's appetite to keep the full specification moving, and it sounds like there is, let's do that.

You can likely close this issue in favor of continuing with https://github.com/w3c/csswg-drafts/issues/4410.

**Comment by bkardell** [comment]
So, @thebabydino is going to be doing some focus on this. She's given more thought to these problems and tried to do more with ranges than pretty much anyone else I can think of, so that's quite good news for openui, imo.  It would be great if we could help direct her on the pragmatic way forward.  She's shared a number of comments on the linked issue already https://github.com/w3c/csswg-drafts/issues/4410#issuecomment-1086891786.  I'm not sure if assembling these all into an explainer (she'd like to not use md, I think) is best or setting right into fleshing out the pages that exist in open ui for 'sliders', or...?  Maybe we can discuss on the call this week

**Comment by lukewarlow** [comment]
Is there anything I can do to help move this along I notice there's research but not an anatomy still. I'm happy to contribute and help come up with an exhaustive anatomy. Might need someone to explain the exact mechanics behind the anatomies.

For example with the below anatomy should progress and thumb (ignoring the names) be inside of track? And should the slot be inside of part or should it wrap it? Select has slots wrap parts but file for example has parts wrap slot?

```js
<AnatomyWrapper>
      <Host name="openui-slider">
        <Part name="track">
          <Slot name="track" />
        </Part>
        <Part name="progress">
          <Slot name="progress" />
        </Part>
        <Part name="thumb">
          <Slot name="thumb" />
        </Part>
      </Host>
    </AnatomyWrapper>
```

**Comment by thebabydino** [comment]
Just commenting to say that, in light of the data I collected from the recent CodePen challenge, I've changed my mind about the anatomy I had in mind prior to this. To exactly what, it's not even clear to me. Every solution comes with its own set of problems.

What I can say is this: if the progress or thumb are inside the track, they will be affected by filter effects, clipping, masking on the track.

For example, we may want to have a filter creating a 3D effect on the track. Let's say we want our track to look like below:

![screenshot of irregular shaped 3D track](https://github.com/openui/open-ui/assets/76854602/9f0d38c1-cec2-46cc-a563-d60ff3716d8e)

We can get this effect with an SVG filter (CodePen [live example](https://codepen.io/thebabydino/pen/RwjzEpm)), but this is also going to affect all descendants - this is why putting the progress and the thumb feels like a bad idea.

Another example would be the case where we want to have a mask on the track, but the thumb should be bigger and expand outside the area retained by the mask:

![screenshot of track with mask and  a bigger thumb](https://user-images.githubusercontent.com/76854602/160440002-19c2e135-5392-456b-831f-a64ad67ba919.png)

However, the track, the progress and the thumb are siblings in the current Firefox implementation and this comes with [major disadvantages](https://gist.github.com/thebabydino/17d690865685abea678791d5307513ab) too.

**Comment by lukewarlow** [comment]
https://github.com/w3c/csswg-drafts/issues/4410#issuecomment-1720875895 it seems the CSSWG resolved on some standard pseudos. It would be good for us to ensure they meet the requirements we've found

**Comment by lukewarlow** [comment]
From a quick glance it seems the main issue with the CSSWG resolved structure for range is that it doesn't support multiple thumbs. How big of a deal do we think that is?

**Comment by mfreed7** [comment]
That'd be good to comment on. Seems important, since many of the examples given in that issue describe multiple thumbs.

**Comment by gregwhitworth** [comment]
> From a quick glance it seems the main issue with the CSSWG resolved structure for range is that it doesn't support multiple thumbs. How big of a deal do we think that is?

I agree that this would be the item that I'd bring up. We don't have psuedo-element short-hands so we can't make `slider-thumb` a shorthand for `slider-thumb-one` and 'slide-thumb-two`.

I will note that I'm personally ok with `slider-thumb` being the name for `slider-thumb-two` (or whatever it's called) because most browsers don't implement this and so it would increase complexity. They could of course implement `slider-thumb-one` without `slider-thumb-two` but that would make it weird for devs to have a name that indicates there should be more than one and it not be available.

**Comment by emilio** [comment]
I wouldn't like to block `::slider-thumb` on a potential `::slider-thumb(n)` / `::slider-thumb-two` or so pseudo... It seems we could implement multiple thumbs in the future if needed. cc @hober, fyi.

**Comment by lukewarlow** [comment]
Yeah I'm not sure it's required to block just something to be aware of to ensure shipping these doesn't preclude it down the line.

**Comment by lukewarlow** [comment]
Various element structures (not sure how much is web exposed)

Blink:

```
<progress>
   <::-webkit-progress-inner-element>
      <::-webkit-progress-bar>
         <::-webkit-progress-value />
      </>
   </>
</progress>
```

```
<meter>
   <::-webkit-meter-inner-element>
      <::-webkit-meter-bar>
         <::-webkit-meter-even-less-good-value />
         or
         <::-webkit-meter-suboptimum-value />
         or
         <::-webkit-meter-optimum-value />
      </>
   </>
</meter>
```

```
<input type="range">
   <::-webkit-slider-container>
      <::-webkit-slider-runnable-track>
         <::-webkit-slider-thumb />
      </>
   </>
</input>
```

WebKit:

```
<progress>
   <::-webkit-progress-inner-element>
      <::-webkit-progress-bar>
         <::-webkit-progress-value />
      </>
   </>
</progress>
```

```
<meter>
   <::-webkit-meter-inner-element>
      <::-webkit-meter-bar>
         <::-webkit-meter-suboptimum-value />
      </>
   </>
</meter>
```

```
<input type="range">
   <::-webkit-slider-container>
      <::-webkit-slider-runnable-track>
         <::-webkit-slider-thumb />
      </>
   </>
</input>
```

Firefox:


```
<progress>
   <::-moz-progress-bar />
</progress>
```

:-moz-meter-sub-sub-optimum::-moz-meter-bar
or
:-moz-meter-sub-optimum::-moz-meter-bar
or
:-moz-meter-optimum::-moz-meter-bar

```
<meter>
   <::-moz-meter-bar />
</progress>
```

```
<input type="range">
   <::-moz-range-track />
   <::-moz-range-progress />
   <::-moz-range-thumb />
</input>
```

I can see it was resolved to apply these to input[type=range], meter and progress. I can I think work out the intention for range inputs

I interpret the below to be the new structure.

```
<input type="range">
   <::slider-track />
   <::slider-fill />
   <::slider-thumb />
</input>
```

The intention for the others is less clear to me?

Would it simply translate to the below?

```
<meter>
   <::slider-track />
   <::slider-fill />
</meter>
```

```
<progress>
   <::slider-track />
   <::slider-fill />
</progress>
```

**Comment by yisibl** [comment]
CSSWG doesn't seem to consider **multiple thumbs**, where do we need to move it forward?
